WIP instrument strchr to use refine-strings

xbauch opened the pull request:
And thus handle the solving via the java string solver. Here we just
1) wrap the argument char-pointer in a refined-string type, i.e. a struct {size,
string}
2) buiild a call `int String.IndexOf(string, char)` to get the offset
3) swap the original call to `lhs=strchr(str,ch)` with `lhs=str+IndexOf(str,ch)`

hannes-steffenhagen-diffblue commented on src/goto-programs/string_instrumentation.cpp:
```diff
@@ -610,6 +622,216 @@ void string_instrumentationt::do_strncmp(
+  const auto apply_index_of = function_application_exprt(
+    symbol_exprt{
+      ID_cprover_string_index_of_func,
+      mathematical_function_typet(std::move(argument_types), index_type())},
```
This `move` doesn't do anything.

hannes-steffenhagen-diffblue committed:
Add c-specific refinement utils

For handling byte-extract expression, string-constants etc., in places that
expect string arrays. Also slightly extend the definition of char-type.

hannes-steffenhagen-diffblue committed:
Extend index-of for c

into a new c-specific axiom-generating function. Basically, the addition encodes
the existence of terminating-zero and the properties this entails.

petr-bauch committed:
Implement strchr in terms of index-of

by
1: associating the `src` with a new infinite array,
2: associating the array with its length variable,
3: applying the c_index_of function
4: translating the returned index to the expected char-pointer/NULL.

petr-bauch committed:
Handle the c-specific cases in string refinement

extracting string arrays from byte-expressions and string constants, hiding the
`symex::args` renaming, etc.

petr-bauch committed:
Add tests for strchr

using the string-refinement solver.

petr-bauch committed:
Small fixes

will be squashed.
